From c7ec952bcd0d97884203403c3bc8ea1f47e8a1c3 Mon Sep 17 00:00:00 2001
From: Stefan Assmann <sassmann@redhat.com>
Date: Tue, 2 Sep 2014 11:37:29 -0400
Subject: [ethernet] i40e: fix Timesync Tx interrupt handler code

Message-id: <1409658042-22689-89-git-send-email-sassmann@redhat.com>
Patchwork-id: 90283
O-Subject: [RHEL7.1 PATCH 088/281] i40e: fix Timesync Tx interrupt handler code
Bugzilla: 1091126
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Jacob Keller <jacob.e.keller@intel.com>

This patch fixes the PTP Tx timestamp interrupt handler. The original
code misinterpreted the interrupt handler design. We were clearing the
ena_mask bit for the Timesync interrupts. This is done to indicate that
the interrupt will be handled in a scheduled work item (instead of
immediately) and that work item is responsible for re-enabling the
interrupts. However, the Tx timestamp was being handled immediately and
nothing was ever re-enabling it. This resulted in a single interrupt
working for the life of the driver.

This patch fixes the issue by instead clearing the bit from icr0 which
is used to indicate that the interrupt was immediately handled and can
be re-enabled right away. This patch also clears up a related issue due
to writing the PRTTSYN_STAT_0 register, which was unintentionally
clearing the cause bits for Timesync interrupts.

Change-ID: I057bd70d53c302f60fab78246989cbdfa469d83b
Signed-off-by: Jacob Keller <jacob.e.keller@intel.com>
Acked-by: Anjali Singhai Jain <anjali.singhai@intel.com>
Acked-by: Shannon Nelson <shannon.nelson@intel.com>
Tested-by: Kavindya Deegala <kavindya.s.deegala@intel.com>
Signed-off-by: Jeff Kirsher <jeffrey.t.kirsher@intel.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
Signed-off-by: Stefan Assmann <sassmann@redhat.com>
(cherry picked from commit cafa1fca9de584dcd920629cf075091a7df64bb0)
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/intel/i40e/i40e_main.c b/drivers/net/ethernet/intel/i40e/i40e_main.c
index e61ea33..bb0a1f2 100644
--- a/drivers/net/ethernet/intel/i40e/i40e_main.c
+++ b/drivers/net/ethernet/intel/i40e/i40e_main.c
@@ -2899,12 +2899,9 @@ static irqreturn_t i40e_intr(int irq, void *data)
   u32 prttsyn_stat = rd32(hw, I40E_PRTTSYN_STAT_0);
 
   if (prttsyn_stat & I40E_PRTTSYN_STAT_0_TXTIME_MASK) {
-   ena_mask &= ~I40E_PFINT_ICR0_ENA_TIMESYNC_MASK;
+   icr0 &= ~I40E_PFINT_ICR0_ENA_TIMESYNC_MASK;
    i40e_ptp_tx_hwtstamp(pf);
-   prttsyn_stat &= ~I40E_PRTTSYN_STAT_0_TXTIME_MASK;
   }
-
-  wr32(hw, I40E_PRTTSYN_STAT_0, prttsyn_stat);
  }
 
  /* If a critical error is pending we have no choice but to reset the
-- 
1.7.1